lazyscribe/repository.py

```diff
@@ -0,0 +1,315 @@
+
+import fsspec
+from attrs import asdict, fields, filters
+
+from lazyscribe._utils import serialize_artifacts
+from lazyscribe.artifacts import _get_handler
+from lazyscribe.artifacts.base import Artifact
+
+LOG = logging.getLogger(__name__)
+
+
+class Repository:
+    """Repository class for holding versioned artifacts.
+
+    Parameters
+    ----------
+    fpath : str | Path, optional (default "repository.json")
+        The location of the repository file. If no repository file exists, this will be the location
+        of the output JSON file when ``save`` is called.
+    mode : {"r", "a", "w", "w+"}, optional (default "w")
+        The mode for opening the repository.
+
+        * ``r``: All existing experiments will be loaded as
+          :py:class:`lazyscribe.experiment.ReadOnlyExperiment` and no new experiments can be logged.
+        * ``a``: All existing experiments will be loaded as
+          :py:class:`lazyscribe.experiment.ReadOnlyExperiment` and new experiments can be added.
+        * ``w``: No existing experiments will be loaded.
+        * ``w+``: All experiments will be loaded in editable mode as
+          :py:class:`lazyscribe.experiment.Experiment`.
+
+    Attributes
+    ----------
+    artifacts : list
+        The list of artifacts in the repository.
+    """
+
+    def __init__(
+        self,
+        fpath: str | Path = "repository.json",
+        mode: Literal["r", "a", "w", "w+"] = "w",
+        **storage_options,
+    ):
+        """Init method."""
+        if isinstance(fpath, str):
+            parsed = urlparse(fpath)
+            self.fpath = Path(parsed.netloc + parsed.path)
+            self.protocol = parsed.scheme or "file"
+        else:
+            self.fpath = fpath
+            self.protocol = "file"
+        self.dir = self.fpath.parent
+        self.storage_options = storage_options
+
+        # If in ``r``, ``a``, or ``w+`` mode, read in the existing repository.
+        self.artifacts: list[Artifact] = []
+        self.snapshot: dict = {}
+        self.fs = fsspec.filesystem(self.protocol, **storage_options)
+
+        if mode not in ("r", "a", "w", "w+"):
+            raise ValueError("Please provide a valid ``mode`` value.")
+        self.mode = mode
+        if mode in ("r", "a", "w+") and self.fs.isfile(self.fpath):
+            self.load()
+
+    def load(self):
+        """Load existing artifacts."""
+        with self.fs.open(self.fpath, "r") as infile:
+            data = json.load(infile)
+
+        artifacts = []
+        for artifact in data:
+            handler_cls = _get_handler(artifact.pop("handler"))
+            created_at = datetime.fromisoformat(artifact.pop("created_at"))
+            artifacts.append(handler_cls.construct(**artifact, created_at=created_at))
+        self.artifacts = artifacts
+
+    def log_artifact(
+        self,
+        name: str,
+        value: Any,
+        handler: str,
+        fname: str | None = None,
+        **kwargs,
+    ):
```

Comment:
Hm... `-> Iterator[Artifact]`?
At least, the user won't know the artifact's `version`/id/object after the `log_artifact` call...

Reply:
Update: and `@contextmanager`.